nexus/src/db/db-macros/src/lookup.rs

```diff
@@ -0,0 +1,649 @@
+fn generate_struct(config: &Config) -> TokenStream {
+    let root_sym = format_ident!("Root");
+    let resource_name = &config.resource.name;
+    let parent_resource_name =
+        config.parent.as_ref().map(|p| &p.name).unwrap_or_else(|| &root_sym);
+    let doc_struct = format!(
+        "Selects a resource of type {} (or any of its children, using the \
+        functions on this struct) for lookup or fetch",
+        resource_name
+    );
+
+    quote! {
+        #[doc = #doc_struct]
+        pub struct #resource_name<'a> {
+            key: Key<'a, #parent_resource_name<'a>>
+        }
+    }
+}
+
+/// Generates the child selectors for this resource
+///
+/// For example, for the "Project" resource with child resources "Instance" and
+/// "Disk", this will generate the `Project::instance_name()` and
+/// `Project::disk_name()` functions.
+fn generate_child_selectors(config: &Config) -> TokenStream {
+    let child_resource_types: Vec<_> =
+        config.child_resources.iter().map(|c| format_ident!("{}", c)).collect();
+    let child_selector_fn_names: Vec<_> = config
+        .child_resources
+        .iter()
+        .map(|c| format_ident!("{}_name", heck::AsSnakeCase(c).to_string()))
+        .collect();
+    let child_selector_fn_docs: Vec<_> = config
+        .child_resources
+        .iter()
+        .map(|child| {
+            format!(
+                "Select a resource of type {} within this {}, \
+                identified by its name",
+                child, config.resource.name,
+            )
+        })
+        .collect();
+
+    quote! {
+        #(
+            #[doc = #child_selector_fn_docs]
+            pub fn #child_selector_fn_names<'b, 'c>(
+                self,
+                name: &'b Name
+            ) -> #child_resource_types<'c>
+            where
+                'a: 'c,
+                'b: 'c,
+            {
+                #child_resource_types {
+                    key: Key::Name(self, name),
+                }
+            }
+        )*
+    }
+}
+
+/// Generates the simple helper functions for this resource
+fn generate_misc_helpers(config: &Config) -> TokenStream {
+    let fleet_name = format_ident!("Fleet");
+    let resource_name = &config.resource.name;
+    let parent_resource_name =
+        config.parent.as_ref().map(|p| &p.name).unwrap_or(&fleet_name);
+
+    // Given a parent authz type, when we want to construct an authz object for
+    // a child resource, there are two different patterns.  We need to pick the
+    // right one.  For "typed" resources (`AuthzKind::Typed`), the parent
+    // resource has a method with the snake case name of the child resource.
+    // For example: `authz_organization.project()`.  For "generic" resources
+    // (`AuthzKind::Generic`), the parent has a function called `child_generic`
+    // that's used to construct all child resources, and there's an extra
+    // `ResourceType` argument to say what resource it is.
```

Comment:
FWIW in order to understand why there are two patterns, I had to read this comment, the one on `enum AuthzKind` on line 42, and then also go look at `authz/api_resources.rs`. It sounds like "typed" is resources that you can attach roles to (org and project) and `generic` is resources that have no permissions of their own, instead inheriting them from their parent. The reason they're called "typed" is you need to use the constructor specific to that resource to get an `Authz` object, whereas for generic children, we don't care what kind of child they are (e.g., disk or instance), we only care that they are children of the parent. (Summarizing to make sure I get it.)

I don't think the comments necessarily need to be changed, but it doesn't seem like there's a single main spot where this is explained. Just giving my experience of where the mental friction is as someone who did not write the code.

https://github.com/oxidecomputer/omicron/blob/52dfddb9f5c0fe4e73910b387ffd5cd3e1414a89/nexus/src/authz/api_resources.rs#L141-L146

Reply:
That's basically right.  Part of what's confusing is that I think of this as two different properties that happen to be correlated:

1. a resource cannot have roles directly attached to it (like anything under a Project)
2. a resource's `authz::` type is constructed by a typed constructor vs. a generic one

Those are the same, in that the typed constructor is used iff the resource can have roles directly attached to it, and that's not really a coincidence.  But it seemed like a pretty big abstraction leak to have the macro know anything about roles.

For what it's worth, another change I want to make soon is to have _all_ of these types use a typed constructor.  I realized recently that there's nothing to stop you from passing an `authz::Instance` where an `authz::Disk` is expected because they're both type aliases for `authz::ProjectChild`.  That's dangerous, so I want to instead use the typed approach (which may mean another macro :-/)

Reply:
Right, it's confusing to use names that are about the code to make a substantive distinction about roles. Typed for everything would make a lot of sense and eliminate a whole thing to understand.

Reply:
I want to echo @david-crespo 's sentiments - I bumped into this when reading the `lookup_resource` arguments, trying to understand when I'd want to supply one AuthzKind vs another, and had to do a lot of digging to understand when to supply one vs the other.

The terms "typed" and "generic" seem to refer a lot more to "how we have implemented these objects" rather than the properties of the objects themselves.

Reply:
Yeah, I get that this is more annoying than we'd like, and as I said, that's why I want to simplify the `authz` types to have a more uniform interface.  Do you have a suggestion for the meantime?

Reply:
I think it’s fine to aim to fix later.